Add unsupported block title translation test

guarani opened the pull request:
## Description


Adds a test to the unsupported block placeholder (a.k.a `missing`) on mobile to ensure the block title is translated when an unsupported (core) block is rendered in the editor. Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/3669.

## How has this been tested?

1. Revert the changes made for the [[RNMobile] Fix missing title for some unsupported blocks](https://github.com/WordPress/gutenberg/pull/31743). The easiest way to do that is `git revert <merge_commit>` (conflicts in the changelog can be marked resolved as-is, since they'll be discarded later).
2. Optional: Verify the translations bug is present by running the packager from within `gutenberg-mobile` (ensuring the changes from Step 1 are present in the Gutenberg submodule), changing the device language to a language other than English, opening the editor from within WPiOS and observing that unsupported blocks **do not** show their translated block name (they show a blank space instead).
3. Run the new test (which should fail, thereby signalling the bug):

```
npm run test-unit:native packages/block-library/src/missing/test/edit-integration.native.js
```
4. Discard the commit produced by the above `git revert` (`git reset --hard <previous_commit>`).
5. Re-run the new test and ensure it passes.

## Types of changes
- New integration tests

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add unsupported block title translation test

guarani commented on packages/block-library/src/missing/test/edit-integration.native.js:
```diff
@@ -0,0 +1,50 @@
+		// jest spy for the _x translation function
+		const _xSpy = jest.spyOn( i18n, '_x' );
+
+		expect( _xSpy ).toHaveBeenCalled();
```
I think this could be improved to ensure that `_x` is indeed called for the purpose of displaying the core block's title in the unsupported block placeholder, and not called for some other purpose (which would also cause this test to pass). 
I'm not sure how to do this yet but will keep looking.

guarani replied:
This difficulty might be related to how `_x` is [exported as a constant](https://github.com/WordPress/gutenberg/blob/b29e3eb11dfac3d19ee5fad52046c2f31822b733/packages/i18n/src/default-i18n.js#L92), because I was getting a `TypeError: Cannot redefine property jest` error here IIRC.

guarani replied:
Also, I'm not sure if it's worth making a separate test to verify that the title is called when the Unsupported Block placeholder is tapped and the bottom sheet is presented.

fluiddot replied:
One of the purposes of integration tests is to query elements in a similar way as the user would do when using the app, for example looking for an element by text or by the a11y label for non-text elements. Due to this reason, I think we should figure out a different approach to verify that the unsupported block title is present and translated.

One idea I'm thinking of is to query by text within the unsupported block, as an example for the Table block, we could try to query if the text `Table` is present.

On the other hand, verifying the translation can be tricky but probably we could set up a locale and include a mocked translation to verify it, wdyt?

**Setup locale code:**
https://github.com/WordPress/gutenberg/blob/2cc7d4f84df4e4dce8c74aee7b32b5b4b979df45/packages/react-native-editor/src/index.js#L120-L151

guarani commented on packages/block-library/src/missing/test/edit-integration.native.js:
```diff
@@ -0,0 +1,50 @@
+ * WordPress dependencies
+ */
+import { getBlockTypes, unregisterBlockType } from '@wordpress/blocks';
+import * as i18n from '@wordpress/i18n';
+jest.mock( '@wordpress/i18n', () => {
+	const actual = jest.requireActual( '@wordpress/i18n' );
+	return {
+		...actual,
+		_x: jest.fn( actual._x ),
+	};
+} );
```
I wanted to verify the `_x` translation function was called while preserving its original implementation, which led me to use `spyOn`. Adding the `mock` here seemed to be the only way to get this to work, after lots of experimentation.

guarani replied:
This difficulty might be related to how `_x` is [exported as a constant](https://github.com/WordPress/gutenberg/blob/b29e3eb11dfac3d19ee5fad52046c2f31822b733/packages/i18n/src/default-i18n.js#L92), because I was getting a `TypeError: Cannot redefine property jest` error here IIRC.

guarani replied:
Marking this as resolved since the new approach doesn't require mocking/spying `_x`.

Commit:
Merge branch 'trunk' into rnmobile/add-unsupported-block-title-translation-test

fluiddot commented:
Continuing with [this comment](https://github.com/WordPress/gutenberg/pull/33340#discussion_r668557304) and since I have some context about the translations due to my work in the translation pipeline improvements, I played around with the i18n package and I think I found a reliable way to test this.

Here is the patch of the changes I did, feel free to edit it:
```patch
diff --git a/packages/block-library/src/missing/test/edit-integration.native.js b/packages/block-library/src/missing/test/edit-integration.native.js
index 4ffa2f8bf0f3945a052f6f83e6d5bc8ac91da1f2..0b2c0d12576182b79eadc6ac599d094434f7818e 100644
--- a/packages/block-library/src/missing/test/edit-integration.native.js
+++ b/packages/block-library/src/missing/test/edit-integration.native.js
@@ -1,20 +1,13 @@
 /**
  * External dependencies
  */
-import { initializeEditor } from 'test/helpers';
+import { initializeEditor, waitFor, within } from 'test/helpers';
 
 /**
  * WordPress dependencies
  */
 import { getBlockTypes, unregisterBlockType } from '@wordpress/blocks';
-import * as i18n from '@wordpress/i18n';
-jest.mock( '@wordpress/i18n', () => {
-	const actual = jest.requireActual( '@wordpress/i18n' );
-	return {
-		...actual,
-		_x: jest.fn( actual._x ),
-	};
-} );
+import { setLocaleData } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
@@ -22,6 +15,8 @@ jest.mock( '@wordpress/i18n', () => {
 import { registerCoreBlocks } from '../..';
 
 beforeAll( () => {
+	setLocaleData( { 'block title\u0004Table': [ 'Tabla' ] } );
+
 	// Register all core blocks
 	registerCoreBlocks();
 } );
@@ -31,6 +26,9 @@ afterAll( () => {
 	getBlockTypes().forEach( ( block ) => {
 		unregisterBlockType( block.name );
 	} );
+
+	// Clean up translations
+	setLocaleData( {} );
 } );
 
 describe( 'Unsupported block', () => {
@@ -38,13 +36,18 @@ describe( 'Unsupported block', () => {
 		const initialHtml = `<!-- wp:table -->
 		 <figure class="wp-block-table"><table><tbody><tr><td>1</td><td>2</td></tr><tr><td>3</td><td>4</td></tr></tbody></table></figure>
 		 <!-- /wp:table -->`;
-		await initializeEditor( {
+		const { getByA11yLabel } = await initializeEditor( {
 			initialHtml,
 		} );
 
-		// jest spy for the _x translation function
-		const _xSpy = jest.spyOn( i18n, '_x' );
+		const missingBlock = await waitFor( () =>
+			getByA11yLabel( /Unsupported Block\. Row 1/ )
+		);
+
+		const translatedTableTitle = within( missingBlock ).getByText(
+			'Tabla'
+		);
 
-		expect( _xSpy ).toHaveBeenCalled();
+		expect( translatedTableTitle ).toBeDefined();
 	} );
 } );
```

### Setting translations
For this purpose, I'm using the function `setLocaleData` from the `i18n` package, which is used to basically set the object with all the translations for the current locale. Here you can see how we use it when initializing the editor:
https://github.com/WordPress/gutenberg/blob/c26c506211be10a82f00e8b5187ed1b369195b9c/packages/react-native-editor/src/index.js#L121-L141

### Block title localization key

The key I used is `block title\u0004Table`, which is composed by the following parts:
- `block title`: this is the context of the key, for the texts defined in the `block.json` file, there's a schema that relates each parameter to its context.
https://github.com/WordPress/gutenberg/blob/b65583def3168fb481b2a776f892f2a38b10a616/packages/blocks/src/api/i18n-block.json#L2

- `\u0004`: this is the [separator character defined in the library](https://github.com/aduth/tannin/blob/3173b45fc788faa118db9c45dd2e0d3bd4b4e6a9/packages/tannin/index.js#L52) used to handle the translation.

- `Table`: this is the text defined in the Table's `block.json` file for the title.
https://github.com/WordPress/gutenberg/blob/c26c506211be10a82f00e8b5187ed1b369195b9c/packages/block-library/src/table/block.json#L4

Commit:
Update unsupported block placeholder test

- Instead of checking if the `_x` translation function was called, we now instead mock the translations and verify that the translated string appears in the UI.
- Add a test to verify translations in the unsupported block bottom sheet title.

guarani commented:
Thank you so much for the detailed response here, @fluiddot! The approach you suggest is great and I've applied the patch to the code and updated the PR. 

I've also added another test to verify that the title e.g. "'Table' is not fully-supported" of the bottom sheet that's shown when the user taps on the help icon is translated.

I'm marking this as ready for review and re-requesting a review from you 🙇.

guarani commented:
Thank you so much for the detailed response here, @fluiddot! The approach you suggest is great and I've applied the patch to the code and updated the PR. 

I've also added another test to verify that the title e.g. "'Table' is not fully-supported" of the bottom sheet that's shown when the user taps on the help icon is translated.

I'm marking this as ready for review and re-requesting a review from you 🙇.

guarani commented:
BTW @fluiddot – I created a new file here, `packages/block-library/src/missing/test/edit-integration.native.js`, because there as already a `test/edit.native.js` file and I thought it might be best not to add integration tests to it since this is a different kind of test. If you think this could be improved some way just let me know 👍.

fluiddot commented:
> BTW @fluiddot – I created a new file here, `packages/block-library/src/missing/test/edit-integration.native.js`, because there as already a `test/edit.native.js` file and I thought it might be best not to add integration tests to it since this is a different kind of test. If you think this could be improved some way just let me know 👍.

It makes sense to me to have them separated but I'm wondering if eventually the original test (which is using the testing library `react-test-renderer`) could be migrated to an integration test. As far as I checked, it's rendering the block and verifying that elements are being rendered as expected which can also be done with the new testing library.

fluiddot commented on packages/block-library/src/missing/test/edit-integration.native.js:
```diff
@@ -0,0 +1,84 @@
+		fireEvent.press( missingBlock );
+
+		const helpButton = await waitFor( () =>
+			getByA11yLabel( /Help button/ )
```
In this case, we could use the text value `Help button` instead of a regular expression as the accessibility label is a static value.

https://github.com/WordPress/gutenberg/blob/5004fe6dc9599f1ff42bef37b08e58dd100fe814/packages/block-library/src/missing/edit.native.js#L89

fluiddot reviewed (approved):
LGTM 🎊 !

I only added a comment with a suggestion but feel free to skip it as it's minor.

Commit:
Tweak element query for clarity

guarani commented:
> LGTM 🎊 !
> 
> I only added a comment with a suggestion but feel free to skip it as it's minor.

Thanks! Good idea, I made the change 👍
